### PR TITLE
feat(smali): improve queries

### PR DESCRIPTION
**Pull request by amaanq**
cc: @ObserverOfTime as he helped me with Smali initially when I was new :) would def like an eyeball or two over the query changes. The grammar was about half-rewritten as it wasn't complete nor correct in its broader statements and organization

Also note that I am a contributor to the sourcehut repo since @yotamN and I agreed to co-maintain the parser over there, no need to worry about that going stale or out of date

### CODE REVIEW

File: queries/smali/highlights.scm [hunk 1]
```diff
@@ -4,52 +4,106 @@
...
 
-(variable) @variable
+(variable) @variable.builtin
+
+(local_directive
+  (identifier) @variable)
+
+; Parameters
+
+(parameter) @parameter ; more like @parameter.builtin but that doesn't exist in nvim-treesitter
```
Comment (ObserverOfTime) [inline queries/smali/highlights.scm:77]: You can still use `@parameter.builtin` if you want.

File: queries/smali/highlights.scm [hunk 2]
```diff
@@ -4,52 +4,106 @@
...
-(array_type
-  "[" @punctuation.bracket)
+((class_identifier) @type.builtin
+  (#lua-match? @type.builtin "^Landroid/"))
+((class_identifier) @type.builtin
+  (#lua-match? @type.builtin "^Lcom/android/"))
+((class_identifier) @type.builtin
+  (#lua-match? @type.builtin "^Ldalvik/"))
+((class_identifier) @type.builtin
+  (#lua-match? @type.builtin "^Ljava/"))
```
Comment (ObserverOfTime) [inline queries/smali/highlights.scm:14]: These should be combined.

File: queries/smali/highlights.scm [hunk 3]
```diff
@@ -4,52 +4,106 @@
...
+  (#lua-match? @conditional "^cmp"))
 
-"=" @operator
+((opcode) @exception
+  (#lua-match? @exception "^throw"))
+
+[
+  "="
+  ".."
+  "@"
```
Comment (ObserverOfTime) [inline queries/smali/highlights.scm:105]: Why is `@` an operator?

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: queries/smali/highlights.scm (hunk 1 above)
Comment (amaanq) [reply queries/smali/highlights.scm:77]: that's not in the list in CONTRIBUTING.md though, is that fine then?

File: queries/smali/highlights.scm (hunk 2 above)
Comment (amaanq) [reply queries/smali/highlights.scm:14]: how so? I tried just stacking each lua-match expression in one but then none would work

File: queries/smali/highlights.scm (hunk 3 above)
Comment (amaanq) [reply queries/smali/highlights.scm:105]:
```smali
    invoke-custom {v0, v1}, call_site_0("doSomething", (LCustom;Ljava/lang/String;)Ljava/lang/String;, "just testing")@LBootstrapLinker;->backwardsLink(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/String;)Ljava/lang/invoke/CallSite;

    invoke-custom {v0, v1}, call_site_2("doSomething", (LCustom;Ljava/lang/String;)Ljava/lang/String;, invoke-static@Lnonsense;->somemethod()V)@LBootstrapLinker;->normalLink(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;)Ljava/lang/invoke/CallSite;
```

very niche statements in Smali

File: queries/smali/highlights.scm (hunk 2 above)
Comment (ObserverOfTime) [reply queries/smali/highlights.scm:14]: This should work.
```clojure
((class_identifier) @type.builtin
  (#vim-match? @type.builtin "^L(android|com/android|dalvik|java)/"))
```

File: queries/smali/highlights.scm (hunk 1 above)
Comment (clason) [reply queries/smali/highlights.scm:77]: Yes. Only top-level captures are enforced; subcaptures can be arbitrary.

Comment (ObserverOfTime) [reply queries/smali/highlights.scm:77]: Yes, it just won't be highlighted any differently by default.

File: queries/smali/highlights.scm (hunk 3 above)
Comment (ObserverOfTime) [reply queries/smali/highlights.scm:105]: Looks more like `@punctuation.delimiter` to me.

File: queries/smali/highlights.scm (hunk 1 above)
Comment (clason) [reply queries/smali/highlights.scm:77]: (careful: we just care about capture name to _allow_ consistent highlighting -- _how_ or even _whether_ something is highlighted completely depends on the colorscheme. we should be careful to always make this distinction.)

File: queries/smali/highlights.scm (hunk 3 above)
Comment (amaanq) [reply queries/smali/highlights.scm:105]: mmm yeah, its to separate targets and their types so that makes more sense to me too then

**Review by ObserverOfTime — APPROVED** [review]

**Pull request merged**